pkg/kube/secretcontroller/secretcontroller_test.go

```diff
@@ -15,126 +15,168 @@
+	mu.Lock()
+	defer mu.Unlock()
+	deleted = id
+	return nil
 }
 
-func verifyControllerCreated(t *testing.T, timeoutName string) {
-	pkgtest.NewEventualOpts(10*time.Millisecond, 5*time.Second).Eventually(t, timeoutName, func() bool {
-		return atomic.LoadInt32(&testCreateControllerCalled) == 1
-	})
+func resetCallbackData() {
+	added = ""
+	updated = ""
+	deleted = ""
 }
 
 func Test_SecretController(t *testing.T) {
+	g := NewWithT(t)
+
 	LoadKubeConfig = mockLoadKubeConfig
 	ValidateClientConfig = mockValidateClientConfig
 	CreateInterfaceFromClusterConfig = mockCreateInterfaceFromClusterConfig
 
 	clientset := fake.NewSimpleClientset()
 
-	// Start the secret controller and sleep to allow secret process to start.
-	err := StartSecretController(
-		clientset, testCreateController, testDeleteController, secretNamespace)
-	if err != nil {
-		t.Fatalf("Could not start secret controller: %v", err)
+	var (
+		secret0                        = makeSecret("s0", "c0", []byte("kubeconfig0-0"))
+		secret0UpdateKubeconfigChanged = makeSecret("s0", "c0", []byte("kubeconfig0-1"))
+		secret0UpdateKubeconfigSame    = makeSecret("s0", "c0", []byte("kubeconfig0-1"))
+		secret1                        = makeSecret("s1", "c1", []byte("kubeconfig1-0"))
+	)
+	secret0UpdateKubeconfigSame.Annotations = map[string]string{"foo": "bar"}
+
+	steps := []struct {
+		// only set one of these per step. The others should be nil.
+		add    *v1.Secret
+		update *v1.Secret
+		delete *v1.Secret
+
+		// only set one of these per step. The others should be empty.
+		wantAdded   string
+		wantUpdated string
+		wantDeleted string
+	}{
+		{add: secret0, wantAdded: "c0"},
+		{update: secret0UpdateKubeconfigChanged, wantUpdated: "c0"},
+		{update: secret0UpdateKubeconfigSame},
+		{add: secret1, wantAdded: "c1"},
+		{delete: secret0, wantDeleted: "c0"},
+		{delete: secret1, wantDeleted: "c1"},
 	}
-	time.Sleep(100 * time.Millisecond)
 
-	// Create the multicluster secret.
-	err = createMultiClusterSecret(clientset)
-	if err != nil {
-		t.Fatalf("Unexpected error on secret create: %v", err)
-	}
-
-	verifyControllerCreated(t, "Create remote secret controller")
-
-	if atomic.LoadInt32(&testDeleteControllerCalled) == 1 {
-		t.Fatalf("Test failed on create secret, delete callback function called")
-	}
-
-	// Reset test variables and delete the multicluster secret.
-	atomic.StoreInt32(&testCreateControllerCalled, 0)
-	atomic.StoreInt32(&testDeleteControllerCalled, 0)
-
-	err = deleteMultiClusterSecret(clientset)
-	if err != nil {
-		t.Fatalf("Unexpected error on secret delete: %v", err)
-	}
+	// Start the secret controller and sleep to allow secret process to start.
+	g.Expect(
+		StartSecretController(clientset, addCallback, updateCallback, deleteCallback, secretNamespace)).
+		Should(Succeed())
 
-	// Test - Verify that the remote controller has been removed.
-	verifyControllerDeleted(t, "delete remote secret controller")
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Why do we need this?

Reply:
Leftover from the previous tests. We don't need it anymore.